### ISSUE TITLE
Conflict when multiple different components on the same flexible

### ISSUE THREAD

**Issue by barnemax-pilotin** [Pilot-in/PiloPress#84]
**Describe the bug**

**1)** When mutiple **different** components are added to the same flexible only the first one (by order in the flexible - i will refer to it as **_FC_** in this thread) will have its data transmitted, not the second one. What is happening is that all the components are receiving the data from **_FC_** (same fields etc) because the data is stocked in the same index : `field_component_wrapper`  (`pilopress/includes/classes/components/class-components.php` l.290) so it return the first results.

My quick fix to this is to add the `$selector`as a suffix to this, so each data is saved by component. I tried to use `get_row_index()` to make possible that every possibility can be stocked and retrieved but this caused error on front. 
So right with : `$field_key = 'field_component_wrapper_' . $selector;` if there are more that one component of the same component with different values there is gonna a problem. **This needs to be improved.**

**2)** Also there is a bug on the admin when multiple components are set, the component selectors retrieves only one of the values from all the components selector (the last one loaded) which cause for all expect the one load to not display the selected value(s) . The value being transmitted by the `prepare_field` (`pilopress/includes/classes/components/class-component-field-type.php` l.115)  is always the one set by the last `load_value` (same file, l.87), all the `load_value`s are being processed before the all the` prepare_field`s

My quick fix to this is to change the `private $initial_value;` to  `private $initial_value = array();` (same file l.19) and all the later uses of `$this->initial_value` to `$this->initial_value[pip_maybe_get( $field, 'key' )]`  so that values are at least the same for the components. The problem with this is if there is two of the same components selector with different values, then maybe using a system based on the index in the flexible could be useful but the `$field` object isn't the same at load_value and prepare_field. **This needs to be improved.**

These appear at lines 90, 94, 118 (2 times), 133.

**To Reproduce**

Put at least 2 different components on the same flexible.

Here is a video of the problem of the selection : https://www.loom.com/share/77527b9e701342a28e6509814df8b684
Here is a video of the layout appareance : https://www.loom.com/share/229c5e3ce64e44f5b67757cf777235c7

**Expected behavior**

All the components should have their own data and the admin showing the selected values properly

**WordPress, Plugins**

WordPress version: 5.5
Pilo'Press version: 0.3.2.10
ACF Pro version: 5.9.1
ACFE version: 0.8.7.2
ACF add-ons: Advanced Custom Fields: Font Awesome (3.1.1)

**Comment by barnemax-pilotin** [comment]
I worked on the 2) issue with viable solution.

The same way of doing as I first said in the original issue but this time mapping each componant field with a unique ID so their values don't get overriden. 

So the here is the list of modifications in `pilopress/includes/classes/components/class-component-field-type.php` : 

### Intializing the variable l.19 : 
`private $initial_value = array();`

### In `load_value` :
At the beginning of `load_value` - creating the unique index:

`$initial_value_index = pip_maybe_get( $field, 'key' ) . '_' . pip_maybe_get( $field, 'name' );`
**Is numeric check :** 
```
if ( is_numeric( $value ) ) {
    $this->initial_value[ $initial_value_index ] = $value;
}
```
**Getting sub_fields:**
`$sub_fields = get_field_objects( pip_maybe_get( $this->initial_value, $initial_value_index ), false );`

### In `prepare_field`:

At the begging of `prepare_field` - basically what I'm doing here is using the field data to match the unique index set in `load_value`

```
$acf_field_prefix = pip_maybe_get( $field, 'prefix' );
preg_match( '~row-(.*?)]~', $acf_field_prefix, $regexed_row_number );
$row_number = $regexed_row_number[1];

//Mapping the index for it to have the same structure as in load_value and format_value
$initial_value_index = pip_maybe_get( $field, 'key' ) . '_' . PIP_Flexible::get_flexible_field_name() . '_' . $row_number . '_' . pip_maybe_get( $field, '_name' );
```

Later in prepare field, when setting the value: 

`$field['value']   = pip_maybe_get( $this->initial_value, $initial_value_index );`

### In `format_value`:

At the begging of `format_value`:

`$initial_value_index = pip_maybe_get( $field, 'key' ) . '_' . pip_maybe_get( $field, 'name' );`

And for the return:

`return get_fields( pip_maybe_get( $this->initial_value, $initial_value_index ), true );`

**Comment by maelys-pilotin** [comment]
Fixed for Pilo'Press 0.4.0
